### PR TITLE
Connection

### PR DESCRIPTION
**Pull request by Pbotsaris**
### Tests
Added unit testing. For now, the code that runs tests is the `main.cpp` of both the server and the client. 

Every time we build and run, tests will run automatically. In the future, we can have a main just for tests or pass in arguments to run tests or not.  You can write tests in any `.cpp` file by including the `doctest.h` header. See an example of tests in the bottom of `connection.cpp` or `text_io.cpp` source files.

For now I added just some very basic testing. Moving forward, we want to expand the unit test coverage on the connection code.

### File Structure
Added a `shared` folder for `src` as well so we can have source files shared by both the server and client.  Note that all connection code is within the `connection` directory for both source files and header files to keep things tidy.

### Connection Implementation
Wrote the underlying socket code to build our clients, servers, etc.  Below are some details about this implementation and its classes.

- `Connection`. the basic connection functionality shared by `ActiveConn` and `PassiveConn` such as creating a socket descriptor, setting socket options, setting the `sockaddr_in` struct etc.
- `ActiveConn` and `PassiveConn`. Specialize connections as either active or passive. Inherits from `Connection` and expands it to have more specialized methods such is `accept_connection` in `PassiveConn` and `connect_socket` in `ActiveConn`.
 
- `ConnectionPoll` used by `PassiveConn`. A passive connection must accept, receive and respond from/to multiple connected client sockets. `ConnectionPoll` handles that. 
- `IOStrategy` is the interface to create in/out data strategies to a connection. This way we can just 'plug in' a strategy to either an active or passive connection in order to send and receive different types of data(audio, video, text etc) on a connection.
- `TextIO` is the only implemented strategy from `IOStrategy` for now. The idea is that we will have other ones in the future such as `AudioIO`, `VideoIO` etc.

For more info on the Strategy design pattern please check this [link](https://refactoring.guru/design-patterns/strategy) or watch this [video](https://www.youtube.com/watch?v=-NCgRD9-C6o&t=251s).

Check [connection.cpp](https://github.com/khalilmasri/Skype/blob/d42a537f6534c5e748dbbd2c41c4427c06c8a063/src/shared/connection/connection.cpp#L76), [active_conn.cpp](https://github.com/khalilmasri/Skype/blob/d42a537f6534c5e748dbbd2c41c4427c06c8a063/src/shared/connection/active_conn.cpp#L34), and [text_io.cpp](https://github.com/khalilmasri/Skype/blob/d42a537f6534c5e748dbbd2c41c4427c06c8a063/src/shared/connection/text_io.cpp#L71)  tests for **examples** on how to use this implementation.

 `connection.cpp` tests only for a very basic connection setup. 

`active_conn.cpp` tests the connection between an active connection to a passive connection.

`text_io.cpp` tests for sending and receiving messages between active and passive connections.

I also put a diagram together(attached and trello) to illustrate the relationship between the classes.
conn_dia

### Moving forward

For now, the `data` argument going to  `receive` and `respond` is only a string.  We need to expand that to be a class/object that can multiple data types such as `audio`, `video`, `text` etc.

### CODE REVIEW

File: src/shared/connection/connection_poll.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  m_poll[m_poll_index].events = 0;
+  m_poll[m_poll_index].events |= POLLIN;
+  m_poll_index++;
+
+  return true;
+}
+
+bool ConnectionPoll::accept_awaits() {
+
+  if (poll(m_poll, 1, m_timeout) > 0) {
```
Comment (khalilmasri) [inline src/shared/connection/connection_poll.cpp:33]: We could add a define or a variable called `zero` and make it `=1` for less confusion

Comment (khalilmasri) [reply src/shared/connection/connection_poll.cpp:33]: Can we also make a variable and then check the return of bigger than 0

```
int res = poll(m_poll, 1, m_timeout);
if( res > 0){
}
```

File: src/shared/connection/passive_conn.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      m_poll.add_socket(res); // add new connections to poll.
+  }
+
+  return valid;
+}
+
+bool PassiveConn::receive(std::string &t_data) {
+  m_client_socket = m_poll.select_socket_with_events();
+
+  bool valid = is_valid(m_client_socket, "Nothing to receive.");
```
Comment (khalilmasri) [inline src/shared/connection/passive_conn.cpp:52]: That seems like a spammy message, we can add another argument that sends a logger level

```
is_valid(m_client_socket, "Nothing to receive.", INFO);
```

**Comment by khalilmasri** [comment]
Seems great, and we can use all of the code in abstraction.

More modifications maybe needed for later, when we experiment more.

I've added some comments, and can we please change the `auto` to the type of the variables instead?

**Review by khalilmasri — CHANGES_REQUESTED** [review]
Seems great, and we can use all of the code in abstraction.

More modifications maybe needed for later, when we experiment more.

I've added some comments, and can we please change the `auto` to the type of the variables instead?

File: src/shared/connection/passive_conn.cpp (hunk 2 above)
Comment (Pbotsaris) [reply src/shared/connection/passive_conn.cpp:52]: @khalilmasri  I am adding a DEBUG log to them so we can turn them on and off.

**Pull request merged**